src/data_source/storage/sql.rs

```diff
@@ -1170,6 +1188,491 @@ where
+                             h.data->>{header_state_commitment_field}"
+                        ),
+                        &[&writer],
+                    )
+                    .await?;
+
+                (query.get(0), writer)
+            }
+            Snapshot::<State>::Index(created) => {
+                let created = created as i64;
+                // Check if header exists
+                // For Snapshot::Commit the entry is already checked when retrieving the block height
+                let row = self.query_one(
+                    "SELECT h.data->>{header_state_commitment_field} as root_commmitment from HEADER where height = $1",
+                    [sql_param(&created)],
+                )
+                .await?;
+
+                (created, row.get(0))
+            }
+        };
+
+        // Get all the nodes in the path to the index.
+        // Order by pos DESC is to return nodes from the leaf to the root
+        let nodes = self
+            .query(
+                &format!(
+                    "SELECT DISTINCT ON (pos) *
+                    FROM {state_type}
+                    WHERE pos @> $1 and created <= $2
+                    ORDER BY pos DESC , created DESC;"
+                ),
+                [sql_param(&LTree::from(path.iter())), sql_param(&created)],
+            )
+            .await?;
+
+        let nodes: Vec<_> = nodes.map(|res| Node::try_from(res?)).try_collect().await?;
+
+        // insert all the hash ids to a hashset which is used a query later
+        // HashSet is used to avoid duplicates
+        let mut hash_ids = HashSet::new();
+        nodes.iter().for_each(|n| {
+            hash_ids.insert(n.hash_id);
+            if let Some(children) = &n.children {
+                hash_ids.extend(children);
+            }
+        });
+
+        // Find all the hash values and created a hashmap
+        // Hashmap will be used to get the hash value of the nodes children and the node itself.
+        let hashes_query = self
+            .query(
+                "SELECT * FROM hash WHERE id = ANY( $1)",
+                [sql_param(&hash_ids.into_iter().collect::<Vec<i32>>())],
+            )
+            .await?;
+
+        let hashes: HashMap<_, _> = hashes_query
+            .map(|row| HashTableRow::try_from(row?).map(|h| (h.id, h.value)))
+            .try_collect()
+            .await?;
+
+        // The merkle tree commitment from the header for the snapshot should match the root node hash in the table
+        // The last node in nodes vector is the root node
+        let root_node = nodes.last().ok_or(QueryError::Error {
+            message: "No node found in the db".to_owned(),
+        })?;
+        // Get the hash value
+        let root_hash_value = hashes.get(&root_node.hash_id).ok_or(QueryError::Error {
+            message: "Hash value for the root node not found".to_owned(),
+        })?;
+        if *root_hash_value != merkle_commitment {
```

Comment:
I think there's something kind of annoying here. The `merkle_commitment` we store in the header is not just the root node hash, it contains some other metadata needed to fully commit to the structure of the tree, like the number of leaves. It's actually a serialization of [this struct](https://github.com/EspressoSystems/jellyfish/blob/main/primitives/src/merkle_tree/internal.rs#L101). So to pull out the root digest for this comparison, we need to parse `merkle_commitment` (it should be a tagged base 64 string) as a `MerkleTreeCommitment` using `serde_json::from_str`. Then you can check like `*root_hash_value == comm.digest()`